src/ray/core_worker/object_recovery_manager.h

```diff
@@ -0,0 +1,150 @@
+    ObjectPinningClientFactoryFn;
+
+typedef std::function<void(const ObjectID &object_id,
+                           const std::vector<rpc::Address> &raylet_locations)>
+    ObjectLookupCallback;
+
+class ObjectRecoveryManager {
+ public:
+  ObjectRecoveryManager(const rpc::Address &rpc_address,
+                        ObjectPinningClientFactoryFn client_factory,
+                        std::shared_ptr<PinObjectsInterface> local_object_pinning_client,
+                        std::function<Status(const ObjectID &object_id,
+                                             const ObjectLookupCallback &callback)>
+                            object_lookup,
+                        std::shared_ptr<TaskResubmissionInterface> task_resubmitter,
+                        std::shared_ptr<ReferenceCounter> reference_counter,
+                        std::shared_ptr<CoreWorkerMemoryStore> in_memory_store,
+                        std::function<void(const ObjectID &object_id, bool pin_object)>
+                            reconstruction_failure_callback,
+                        bool lineage_reconstruction_enabled)
+      : rpc_address_(rpc_address),
+        client_factory_(client_factory),
+        local_object_pinning_client_(local_object_pinning_client),
+        object_lookup_(object_lookup),
+        task_resubmitter_(task_resubmitter),
+        reference_counter_(reference_counter),
+        in_memory_store_(in_memory_store),
+        reconstruction_failure_callback_(reconstruction_failure_callback),
+        lineage_reconstruction_enabled_(lineage_reconstruction_enabled) {}
+
+  /// Recover an object that was stored in plasma. This will only succeed for
+  /// objects that are lost from memory and that this process owns (returns
+  /// Status::Invalid if false).  This method is idempotent for overlapping
+  /// recovery operations on the same object. This class will guarantee that
+  /// each recovery operation ends in either success (by storing a new value
+  /// for the object in the direct memory/plasma store) or failure (by calling
+  /// the given reconstruction failure callback).
+  ///
+  /// Algorithm:
+  /// 1. Check that the object is missing from the direct memory store and that
+  /// we own the object. If either is false, then fail the recovery operation.
+  /// 2. Look up the object in the global directory to check for other
+  /// locations of the object. If another location exists, attempt to pin it.
+  /// If the pinning is successful, then mark the recovery as a success by
+  /// storing a new value for the object in the direct memory store.
+  /// 3. If pinning fails at all locations for the object (or there are no
+  /// locations), attempt to reconstruct the object by resubmitting the task
+  /// that created the object. If the task resubmission fails, then the
+  /// fail the recovery operation.
+  /// 4. If task resubmission succeeds, recursively attempt to recover any
```

Comment:
It's nice to resubmit the task first because then we can be sure that the reconstructing objects will get pinned by their ref counts. Maybe this is okay even in the reverse order, but it seemed more complicated to convince myself of the correctness.